### PR TITLE
HW #3 (Tests)

### PR DESCRIPTION
**Pull request by eluyema**
- update jest configurations
- add unit tests
- add jest configuration for integration tests
- add docker-compose file for test infrastructure
- add integration tests

### CODE REVIEW

**Comment by cholombytko** [comment]
All your implemented tests look good to me and cover every important aspect of your application. Well done!

File: .env.integration-test [hunk 1]
```diff
@@ -0,0 +1,16 @@
+SERVER_PORT=3000
+SERVER_HOST=0.0.0.0
+
+DATABASE_URL=postgres://user:password@localhost:5438/mydatabase
+DATABASE_USER=user
+DATABASE_PASSWORD=password
+DATABASE_DB=mydatabase
+# test credentials for integration test
+EXCHANGE_API_URL=https://openexchangerates.org/api/latest.json?app_id=5efbcc4908bf449293137c5a83c1f7b2
```
Comment (Larisa-Antsifrova) [inline .env.integration-test:9]: Is that some app_id for testing?)

I wonder if that is safe to publish your own api keys to github)

File: src/modules/exchange-rate/application/integration-test/fetch-exchange-rate.application.int-spec.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  it('class should exist', () => {
+    expect(application).toBeDefined();
+  });
+
+  it('should return exchange rate', async () => {
+    const currentTime = new Date();
+    const exchangeRate = await application.execute();
+    expect(exchangeRate).toBeInstanceOf(ExchangeRate);
+    expect(exchangeRate.rate).toBeGreaterThan(0);
+    expect(exchangeRate.date.getTime()).toBeGreaterThanOrEqual(
```
Comment (Larisa-Antsifrova) [inline src/modules/exchange-rate/application/integration-test/fetch-exchange-rate.application.int-spec.ts:63]: Could you please expand on the logic behind this condition?)

File: src/modules/mailer/application/send-email.application.spec.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  beforeEach(async () => {
+    emailService = new TestMailerService();
+    application = new SendEmailApplicationImpl(emailService);
+  });
+
+  it('should be defined', () => {
+    expect(application).toBeDefined();
+  });
+
+  it('should be defined', async () => {
```
Comment (Larisa-Antsifrova) [inline src/modules/mailer/application/send-email.application.spec.ts]: Probably the test title here should be different)

File: src/modules/mailer/domain/factories/email.factory.spec.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      subject: 'Good news',
+      html: '<b>The weather is good today!</b>',
+    };
+    const email = EmailFactory.create(data);
+    expect(email).toBeDefined();
+    expect(email.html).toEqual(data.html);
+    expect(email.to).toEqual(data.to);
+    expect(email.subject).toEqual(data.subject);
+    expect(email.text).toBeUndefined();
+    expect(email).toBeInstanceOf(Email);
```
Comment (Larisa-Antsifrova) [inline src/modules/mailer/domain/factories/email.factory.spec.ts:52]: How does this test specifically test several e-mails?)

**Comment by Larisa-Antsifrova** [comment]
Finally, I always wonder how developers come up with the test cases they believe should be covered) I would appreciate it if you shared some of your approaches)

File: .env.integration-test (hunk 1 above)
Comment (eluyema) [reply .env.integration-test:9]: Yes, it's id for testing )

File: src/modules/mailer/domain/factories/email.factory.spec.ts (hunk 4 above)
Comment (eluyema) [reply src/modules/mailer/domain/factories/email.factory.spec.ts:52]: This line check it: expect(email.to).toEqual(data.to);

**Pull request merged**